src/plugins/kibana_react/public/context/context.tsx

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import * as React from 'react';
+import { KibanaReactContext, KibanaReactContextValue, KibanaServices } from './types';
+import { createReactOverlays } from '../overlays';
+import { createNotifications } from '../notifications';
+
+const defaultContextValue = {
+  services: {},
+  overlays: undefined,
+  notifications: undefined,
+};
+
+export const context = React.createContext<KibanaReactContextValue<KibanaServices>>(
+  defaultContextValue
+);
+
+export const useKibana = <Extra extends object = {}>(): KibanaReactContextValue<
+  KibanaServices & Extra
+> =>
+  React.useContext((context as unknown) as React.Context<
+    KibanaReactContextValue<KibanaServices & Extra>
+  >);
+
+export const withKibana = <Props extends { kibana: KibanaReactContextValue<any> }>(
+  type: React.ComponentType<Props>
+): React.FC<Omit<Props, 'kibana'>> => {
+  const enhancedType: React.FC<Omit<Props, 'kibana'>> = (props: Omit<Props, 'kibana'>) => {
+    const kibana = useKibana();
+    return React.createElement(type, { ...props, kibana } as Props);
+  };
+  return enhancedType;
+};
+
+export const UseKibana: React.FC<{
+  children: (kibana: KibanaReactContextValue<any>) => React.ReactNode;
+}> = ({ children }) => <>{children(useKibana())}</>;
+
+export const createContext = <Services extends KibanaServices>(
```

Comment:
My initial reaction to this is that it's confusing, it _looks_ like it's just gonna proxy in the React `createContext` method for no reason, or worse, people just think that's exactly what it is once it's been imported. 

Personally I'd prefer to have a `KibanaContextProvider` exported ready to wrap an application, something like:
```
import { KibanaContextProvider } from 'react-kibana';

// ...
// get core from lifecycle method...
<KibanaContextProvider core={core}>
  <MyApp />
</KibanaContextProvider>
```
In most cases that's going to be the path of least resistance, I think, and prevents people having to mess with creating a context. You could export this function as `createKibanaContext` for power users who want to do something more specialized like you have in your examples, with limiting what parts of core are included, or multiple instances of the provider required, etc.

Everything else about this looks really exciting though! Thanks for the work here :)

Reply:
I've added `<KibanaContextProvider>`.